napari/layers/_layer_actions.py

```diff
@@ -50,13 +50,16 @@ def _split_rgb(ll: LayerList):
 def _convert(ll: LayerList, type_: str):
     from napari.layers import Shapes
 
+    float_dtypes = [np.float64, np.float32, np.float16]
     for lay in list(ll.selection):
         idx = ll.index(lay)
         ll.pop(idx)
         if isinstance(lay, Shapes) and type_ == 'labels':
             data = lay.to_labels()
+        elif lay.data.dtype in float_dtypes and type_ == 'labels':
+            data = lay.data.astype(int, copy=False)
         else:
-            data = lay.data.astype(int) if type_ == 'labels' else lay.data
+            data = lay.data
```

Comment:
What's truly wild is that when you return `lay.data` from a boolean image array, the output in the layer data is `int8` instead of `bool`. I don't see where that conversion happens, it must be somewhere after this callback function runs. 

Can someone help explain how I can see the full path the data passes through? 

## Minimal reproducible example

```
conda activate napari-dev
cd GitHub/napari
gh pr checkout 5542
```

```python
import napari
import numpy as np

boolarr = np.array([[1,0,1],[0,1,0],[1,0,1]]).astype(bool)
assert boolarr.dtype == bool

viewer = napari.Viewer()
viewer.add_image(boolarr)
assert viewer.layers[-1].data.dtype == bool

# In the napari viewer, right click the image layer and select "Convert to Labels"
print(viewer.layers[-1].data.dtype)  # surprisingly, returns int8 instead of bool dtype. Where is that conversion happening?
assert viewer.layers[-1].data.dtype == bool  # this fails!
```

Reply:
the change is done here https://github.com/napari/napari/blob/8f79d51e6fbd1031e4f0ed31cd60c3c3333850e2/napari/layers/labels/labels.py#L279